clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -0,0 +1,221 @@
+            kind: ExprKind::Block(block, _),
+            ..
+        }) | ExprKind::Break(_, Some(Expr {
+            kind: ExprKind::Block(block, _),
+            ..
+                                })) | ExprKind::Ret(Some(Expr {
+                               kind: ExprKind::Block(block, _),
+                               ..
+                           })) | ExprKind::Repeat(Expr {
+                             kind: ExprKind::Block(block, _),
+                             ..
+                         }, _) | ExprKind::Yield(Expr{
+                            kind: ExprKind::Block(block, _),
+                            ..
+                        }, YieldSource::Yield) => Some((block, 1)),
+        _ => None
+    }
+}
+
+fn find_candidate(cx: &LateContext<'_>, span: Span, enclosing_hir_id: HirId) {
+    if let Some(error_kind) = block_has_safety_comment(cx.tcx, enclosing_hir_id, span) {
+        match error_kind {
+            ErrorKind::NoComment => {
+                let block_indent = indent_of(cx, span);
+                let suggestion = format!("// Safety: ...\n{}", snippet(cx, span, ".."));
+
+                span_lint_and_sugg(
+                    cx,
+                    UNDOCUMENTED_UNSAFE_BLOCKS,
+                    span,
+                    "unsafe block missing a safety comment",
+                    "consider adding a safety comment",
+                    reindent_multiline(Cow::Borrowed(&suggestion), true, block_indent).to_string(),
+                    Applicability::HasPlaceholders,
+                );
+            }
+            ErrorKind::Placeholder => {
+                span_lint_and_help(
+                    cx,
+                    UNDOCUMENTED_UNSAFE_BLOCKS,
+                    span,
+                    "unsafe block has safety comment placeholder",
+                    None,
+                    "consider replacing the placeholder with a safety comment"
+                );
+            }
+        }
+    }
+}
+
+fn block_has_safety_comment(tcx: TyCtxt<'_>, enclosing_hir_id: HirId, block_span: Span) -> Option<ErrorKind> {
+    let map = tcx.hir();
+    let source_map = tcx.sess.source_map();
+
+    let enclosing_scope_span = map.opt_span(enclosing_hir_id)?;
+    let between_span = enclosing_scope_span.to(block_span);
+
+    let file_name = source_map.span_to_filename(between_span);
+    let source_file = source_map.get_source_file(&file_name)?;
+
+    let lex_start = (between_span.lo().0 + 1) as usize;
+    let src_str = source_file.src.as_ref()?[lex_start..between_span.hi().0 as usize].to_string();
+
+    let mut pos = 0;
+    let mut comment = false;
+
+    for token in rustc_lexer::tokenize(&src_str) {
+        match token.kind {
+            TokenKind::LineComment { doc_style: None }
+            | TokenKind::BlockComment { doc_style: None, terminated: true }  => {
+                let comment_str = src_str[pos + 2..pos + token.len].to_ascii_uppercase();
```

Comment:
What happens, if the comment is not ASCII? Please add a test, where the comment contains non-ascii characters.